pkg/iam/iam.go

```diff
@@ -0,0 +1,409 @@
+// Package iam checks that the AWS user currently logged in
+// has sufficient permissions to run the installer for both
+// `create` and `destroy` operations.
+package iam
+
+import (
+	"encoding/json"
+	"net/url"
+	"regexp"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/sirupsen/logrus"
+)
+
+// AccountDetails stores the AWS account-wide groups and policies.
+type AccountDetails struct {
+	users    []*iam.UserDetail
+	groups   []*iam.GroupDetail
+	policies []*iam.ManagedPolicyDetail
+}
+
+// actions is a slice containing the names of AWS IAM permissions
```

Comment:
can we make sure our ci credentials for aws have exactly these permissions (and no more)?

Reply:
> can we make sure our ci credentials for aws have exactly these permissions (and no more)?

Is this a blocker for this PR?  Currently our CI robot has very broad permissions:

```console
$ AWS_PROFILE=ci aws iam list-groups-for-user --user-name origin-ci-robot-provision
{
    "Groups": [
        {
            "Path": "/", 
            "CreateDate": "2018-02-13T16:07:53Z", 
            "GroupId": "AGPAJF25YZOW6F3SY4FYK", 
            "Arn": "arn:aws:iam::460538899914:group/RobotBuilders", 
            "GroupName": "RobotBuilders"
        }
    ]
}
$ AWS_PROFILE=ci aws iam list-attached-group-policies --group-name RobotBuilders
{
    "AttachedPolicies": [
        {
            "PolicyName": "AmazonEC2FullAccess", 
            "PolicyArn": "arn:aws:iam::aws:policy/AmazonEC2FullAccess"
        }, 
        {
            "PolicyName": "IAMFullAccess", 
            "PolicyArn": "arn:aws:iam::aws:policy/IAMFullAccess"
        }, 
        {
            "PolicyName": "AmazonS3FullAccess", 
            "PolicyArn": "arn:aws:iam::aws:policy/AmazonS3FullAccess"
        }, 
        {
            "PolicyName": "ELBServerCert", 
            "PolicyArn": "arn:aws:iam::460538899914:policy/ELBServerCert"
        }, 
        {
            "PolicyName": "AmazonRoute53FullAccess", 
            "PolicyArn": "arn:aws:iam::aws:policy/AmazonRoute53FullAccess"
        }
    ]
}
```

Reply:
(Just FYI, I edited the above comment to redact information. I think having our AWS account numbers displayed publicly is a security risk).

Reply:
> can we make sure our ci credentials for aws have exactly these permissions (and no more)?

In https://jira.coreos.com/browse/CORS-800 we'll create a user or profile or something (undecided) that will be limited strictly to the permissions listed in this PR. That might achieve what you're looking for. It'll narrow the permissions available to the installer at run time, and help keep our permissions list up-to-date.

Reply:
Yeah that seems more appropriate.